idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -369,11 +374,33 @@ internal object FirReferenceResolveHelper {
         val classId = fir.classId ?: return emptyList()
 
         var parent = expression.parent as? KtDotQualifiedExpression
+        val referencedName = expression.getReferencedName()
+        if (parent != null && expression == parent.receiverExpression && classId.packageFqName.startsWith(Name.identifier(referencedName))) {
+            // the expression references a package under the root
+            return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(FqName(referencedName)))
+        }
         // Distinguish A.foo() from A(.Companion).foo()
         // Make expression.parent as? KtDotQualifiedExpression local function
         while (parent != null) {
             val selectorExpression = parent.selectorExpression ?: break
             if (selectorExpression === expression) {
+                val selectedFqNameSegments = parent.fqNameSegments()
+                if (selectedFqNameSegments != null) {
+                    val fqNameString = classId.asString()
+                    val selectedFqNameSegmentsString = selectedFqNameSegments.joinToString(".")
+                    if (fqNameString.replace('/', '.').startsWith(selectedFqNameSegmentsString)) {
+                        val selectedFqNameString = fqNameString.substring(0, selectedFqNameSegmentsString.length)
+                        if (classId.packageFqName.asString().length >= selectedFqNameString.length) {
+                            // current seletion is a package
+                            return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(FqName.fromSegments(selectedFqNameSegments)))
+                        } else {
+                            // current seletion is a class
+                            return listOfNotNull(
+                                ClassId.fromString(selectedFqNameString).toTargetPsi(session, symbolBuilder)
```

Comment:
How would it work with references to local classes?

Reply:
It doesn't work. In fact, even FIR doesn't work (https://youtrack.jetbrains.com/issue/KT-47135) 🙄